Update added types and methoddefs on ApplyMetadata

davmason opened the pull request:
This change is necessary to allow metadata rewriting at ReJIT time instead of at module load time

noahfalk commented on src/vm/ceeload.cpp:
```diff
@@ -192,6 +192,35 @@ BOOL Module::SetTransientFlagInterlocked(DWORD dwFlag)
 }
 
 #if PROFILING_SUPPORTED 
+void Module::UpdateNewlyAddedTypes()
```
You should also open an issue in the docs repo to track updating the docs to match the new behavior.

davmason replied:
Opened https://github.com/dotnet/docs/pull/10510

davmason commented on src/vm/ceeload.h:
```diff
@@ -1634,6 +1634,11 @@ class Module
     CORCOMPILE_METHOD_PROFILE_LIST *m_methodProfileList;
 
+#if PROFILING_SUPPORTED 
+    DWORD                   m_dwTypeCount;
```
Done, thanks

davmason commented on src/vm/proftoeeinterfaceimpl.cpp:
```diff
@@ -9661,12 +9661,12 @@ HRESULT ProfToEEInterfaceImpl::ApplyMetaData(
     CONTRACTL_END;
 
-    PROFILER_TO_CLR_ENTRYPOINT_SYNC_EX(kP2EEAllowableAfterAttach, (LF_CORPROF, LL_INFO1000, "**PROF: ApplyMetaData.\n"));
+    PROFILER_TO_CLR_ENTRYPOINT_SYNC_EX(kP2EEAllowableAfterAttach | kP2EETriggers, (LF_CORPROF, LL_INFO1000, "**PROF: ApplyMetaData.\n"));
```
Done

noahfalk commented on Documentation/Profiling/Profiler Breaking Changes.md:
```diff
@@ -0,0 +1,6 @@
+Over time we will need to modify the Profiler APIs, this document will serve as a record of any breaking changes.
+
+1. Code Versioning introduced changes documented [here](../design-docs/code-versioning-profiler-breaking-changes.md)
+2. The work to allow adding new types and methods after module load means ICorProfilerInfo7::ApplyMetadata will now potentially trigger a GC, and will not be callable in situations where a GC can not happen (for example  ICorProfilerCallback::RootReferences).
```
Profiler vendors may ask us to explain in more detail, but I'm fine keeping this lightweight as you have it now and waiting for them to ask.

noahfalk replied:
When you ping the profiler announce thread, I would encourage them to reach out to us if they have questions and concerns though. I think as long as we are open everyone will be happy.

davmason commented:
@dotnet-bot test coreclr-ci (Test Pri0 Windows_NT x86 checked)

davmason commented:
@noahfalk  can you take a look at the changes I just added? I added a fix for #18536 (a port of your fix from https://github.com/noahfalk/coreclr/commit/5f37ab7686c97b69d86365a3f4413fb77bcefff4 plus making it work for exported types as well)

davmason commented:
@noahfalk I need to take a second look at this. I don't think I looked at LazyPopulateCaseSensitiveHashTables close enough. I don't think it does what I want and isn't safe to call repeatedly.

noahfalk commented on src/vm/ceeload.h:
```diff
@@ -1634,6 +1634,11 @@ class Module
     CORCOMPILE_METHOD_PROFILE_LIST *m_methodProfileList;
 
+#if PROFILING_SUPPORTED 
+    DWORD                   m_dwTypeCount;
```
I forgot before, you actually want PROFILING_SUPPORTED_DATA instead of PROFILING_SUPPORTED. The issue is that the DAC build does not define PROFILING_SUPPORTED, but if these fields are missing then its layout computations for Module will be wrong. You won't see a bug right now, but if we added new fields to the module after these and used them in the debugger you would see it.

davmason replied:
Thanks!

davmason commented:
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test
@dotnet-bot test Ubuntu x64 Formatting

davmason commented:
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Checked Innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Checked crossgen_comparison Build and Test
@dotnet-bot test Ubuntu arm Cross Checked no_tiered_compilation_innerloop Build and Test
@dotnet-bot test Ubuntu arm Cross Release crossgen_comparison Build and Test
@dotnet-bot test Ubuntu x64 Checked CoreFX Tests

davmason commented:
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

davmason commented:
I don’t know what’s up with the CI, but all these legs were passing before I reset them. I’m going to merge now.
